README
NAME
    POEx::Weather::OpenWeatherMap - POE-enabled OpenWeatherMap client

SYNOPSIS
      use POE;
      use POEx::Weather::OpenWeatherMap;

      # An API key can be obtained (free) at http://www.openweathermap.org:
      my $api_key = 'foo';

      POE::Session->create(
        package_states => [
          main => [qw/
            _start
        
            pwx_error
            pwx_weather
            pwx_forecast
          /],
        ],
      );

      sub _start {
        my ($kernel, $heap) = @_[KERNEL, HEAP];
    
        # Create, store, and start emitter:
        my $wx = POEx::Weather::OpenWeatherMap->new(
          api_key      => $api_key,
          event_prefix => 'pwx_',
        );

        $heap->{wx} = $wx;
        $wx->start;

        ## An example request:
        $wx->get_weather(
          location => 'Manchester, NH',
          tag      => 'mytag',
        );
      }

      sub pwx_error {
        my $err = $_[ARG0];
        my $status  = $err->status;
        my $request = $err->request;
        # ... do something with error ...
        warn "Error! ($status)";
      }

      sub pwx_weather {
        my $result = $_[ARG0];

        my $tag = $result->request->tag;

        my $place = $result->name;

        my $tempf = $result->temp_f;
        my $conditions = $result->conditions_verbose;
        # (see Weather::OpenWeatherMap::Result::Current for a method list)
        # ...
      }

      sub pwx_forecast {
        my $result = $_[ARG0];

        my $place = $result->name;

        my $itr = $result->iter;
        while (my $day = $itr->()) {
          my $date = $day->dt->mdy;
          my $temp_hi = $day->temp_max_f;
          my $temp_lo = $day->temp_min_f;
          # (see Weather::OpenWeatherMap::Result::Forecast)
          # ...
        }
      }

      POE::Kernel->run;

DESCRIPTION
    A POE-enabled interface to OpenWeatherMap
    (<http://www.openweathermap.org>), providing an object-oriented
    asynchronous interface to current & forecast weather conditions for a
    given city, latitude/longitude, or OpenWeatherMap city code.

    This is really just an asynchronous counterpart to
    Weather::OpenWeatherMap; look there for documentation regarding Request
    & Result objects.

    This an event emitter that consumes MooX::Role::POE::Emitter; look there
    for documentation on composed methods. See
    <http://www.openweathermap.org> for more on OpenWeatherMap itself.

  ATTRIBUTES
   api_key
    Your OpenWeatherMap <http://www.openweathermap.org> API key.

    (See <http://www.openweathermap.org/api> to register for free.)

   cache
    A boolean value indicating whether successful results should be cached
    to disk via Weather::OpenWeatherMap::Cache.

    Defaults to false. This may change in a future release.

   cache_dir
    The directory in which cache files are saved. The default may be fine;
    see Weather::OpenWeatherMap::Cache.

   cache_expiry
    The duration (in seconds) for which cache files are considered valid.
    The default may be fine; see Weather::OpenWeatherMap::Cache.

  METHODS
   start
    Start our session.

    Must be called before events will be received or emitted.

   stop
    Stop our session, shutting down the emitter and user agent (which will
    cancel pending requests).

   get_weather
      $wx->get_weather(
        # 'location =>' is mandatory.
        #  These are all valid location strings:
        #  By name:
        #   'Manchester, NH'
        #   'London, UK'
        #  By OpenWeatherMap city code:
        #   5089178
        #  By latitude/longitude:
        #   'lat 42, long -71'
        location => 'Manchester, NH',

        # Set 'forecast => 1' to get the forecast,
        # omit or set to false for current weather:
        forecast => 1,

        # If 'forecast' is true, you can specify the number of days to fetch
        # (up to 14):
        days => 3,

        # Optional tag for identifying the response to this request:
        tag  => 'foo',
      );

    Request a weather report for the given "location =>".

    The location can be a 'City, State' or 'City, Country' string, an
    OpenWeatherMap <http://www.openweathermap.org/> city code, or a 'lat X,
    long Y' string.

    Requests the current weather by default (see
    Weather::OpenWeatherMap::Request::Current).

    If passed "forecast => 1", requests a weather forecast (see
    Weather::OpenWeatherMap::Request::Forecast), in which case "days =>
    $count" can be specified (up to 14).

    If passed "find => 1", requests search results for a given location name
    or latitude & longitude; see Weather::OpenWeatherMap::Request::Find.

    An optional "tag =>" can be specified to identify the response when it
    comes in.

    Any extra arguments are passed to the constructor for the appropriate
    Request subclass; see Weather::OpenWeatherMap::Request.

    The request is made asynchronously and a response (or error) emitted
    when it is available; see "EMITTED EVENTS". There is no useful return
    value.

  RECEIVED EVENTS
   get_weather
      $poe_kernel->post( $wx->session_id =>
        get_weather =>
          location => 'Manchester, NH',
          tag      => 'foo',
      );

    POE interface to the "get_weather" method; see "METHODS" for available
    options.

  EMITTED EVENTS
   error
    Emitted when an error occurs; this may be an internal error, an HTTP
    error, or an error reported by the OpenWeatherMap API.

    $_[ARG0] is a Weather::OpenWeatherMap::Error object.

   weather
    Emitted when a request for the current weather has been successfully
    processed.

    $_[ARG0] is a Weather::OpenWeatherMap::Result::Current object; see that
    module's documentation for details on retrieving weather information.

   forecast
    Emitted when a request for a weather forecast has been successfully
    processed.

    $_[ARG0] is a Weather::OpenWeatherMap::Result::Forecast object; see that
    module's documentation for details on retrieving per-day forecasts
    (Weather::OpenWeatherMap::Result::Forecast::Day objects).

SEE ALSO
    Weather::OpenWeatherMap

    Weather::OpenWeatherMap::Error

    Weather::OpenWeatherMap::Result

    Weather::OpenWeatherMap::Result::Current

    Weather::OpenWeatherMap::Result::Forecast

    Weather::OpenWeatherMap::Request

    Weather::OpenWeatherMap::Request::Current

    Weather::OpenWeatherMap::Request::Forecast

    The "examples/" directory of this distribution.

AUTHOR
    Jon Portnoy <avenj@cobaltirc.org>



Changes
Revision history for Perl module POEx-Weather-OpenWeatherMap

0.002001 2014-08-16

  - Support 'find' requests

0.001005 2014-05-09

  - Add cache support via Weather::OpenWeatherMap::Cache

0.001004 2014-04-25

  - use Defaults::Modern;

0.001003 2014-04-24

  - POD, test tweaks

0.001002 2014-04-23

  - Drop some deprecated leftovers ('_in_shutdown' attr)

  - Packaging fixes; missing MooX::Role::POE::Emitter dependency

0.001001 2014-04-23

  - Initial release


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.020.
use strict;
use warnings;

use 5.016000;

use ExtUtils::MakeMaker ;



my %WriteMakefileArgs = (
  "ABSTRACT" => "POE-enabled OpenWeatherMap client",
  "AUTHOR" => "Jon Portnoy <avenj\@cobaltirc.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "POEx-Weather-OpenWeatherMap",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "POEx::Weather::OpenWeatherMap",
  "PREREQ_PM" => {
    "Defaults::Modern" => 0,
    "Moo" => 1,
    "MooX::Role::POE::Emitter" => 0,
    "MooX::late" => "0.014",
    "POE" => 1,
    "POE::Component::Client::HTTP" => 0,
    "Weather::OpenWeatherMap" => "0.001005"
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "List::Util" => 0,
    "Scalar::Util" => 0,
    "Test::More" => "0.88",
    "version" => "0.77"
  },
  "VERSION" => "0.002001",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Defaults::Modern" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "List::Util" => 0,
  "Moo" => 1,
  "MooX::Role::POE::Emitter" => 0,
  "MooX::late" => "0.014",
  "POE" => 1,
  "POE::Component::Client::HTTP" => 0,
  "Scalar::Util" => 0,
  "Test::More" => "0.88",
  "Weather::OpenWeatherMap" => "0.001005",
  "version" => "0.77"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





examples/weatherbot.pl
#!/usr/bin/env perl

use v5.10;
use strictures 1;

use List::Objects::WithUtils;
use List::Objects::Types -all;

use POE;
use POEx::Weather::OpenWeatherMap;
use POEx::IRC::Client::Lite;


use Getopt::Long;
my $Opts = +{
  nickname => 'Aurae',
  username => 'aurae',
  server   => undef,
  api_key  => undef,
  channels => '',
  cmd      => '.wx',

  help => sub {
    say $_ for (
      "Usage:",
      "",
      "  --api-key=KEY",
      "",
      "  --nickname=NICKNAME",
      "  --username=USERNAME",
      "  --server=ADDR",
      "  --channels=CHAN[,CHAN ..]",
      "  --cmd=CMD",
    );
    exit 0
  },
};
GetOptions( $Opts,
  'nickname=s',
  'username=s',
  'server=s',
  'api_key=s',
  'channels=s',
  'cmd=s',

  'help',
);

sub getopts { 
  unless (is_ArrayObj $Opts->{channels}) {
    $Opts->{channels} = array(split /,/, $Opts->{channels})
  }
  state $argv = hash(%$Opts)->inflate 
}


POE::Session->create(
  package_states => [
    main => [qw/
      _start

      pxi_irc_001
      pxi_irc_public_msg
      
      pwx_error
      pwx_weather
      pwx_forecast
    /],
  ],
);

sub _start {
  $_[HEAP]->{irc} = POEx::IRC::Client::Lite->new(
    event_prefix => 'pxi_',
    server   => getopts->server,
    nick     => getopts->nickname,
    username => getopts->username,
  );
  $_[HEAP]->{irc}->connect;

  $_[HEAP]->{wx} = POEx::Weather::OpenWeatherMap->new(
    event_prefix => 'pwx_',
    api_key => getopts->api_key,
  );
  $_[HEAP]->{wx}->start;

  $_[HEAP] = hash(%{ $_[HEAP] })->inflate;
}



sub pxi_irc_001 {
  $_[HEAP]->irc->join( getopts->channels->all )
}

sub pxi_irc_public_msg {
  my $event = $_[ARG0];
  my ($target, $string) = @{ $event->params };
  
  my $cmd = getopts->cmd;
  if ( index($string, "$cmd ") == 0 ) {
    my ($location, $fcast);
    if (index($string, 'forecast') == length($cmd)+1) {
      $location = substr $string, length($cmd) + length('forecast') + 2;
      $fcast++
    }

    $location ||= substr $string, length("$cmd ");

    $_[HEAP]->wx->get_weather(
      location => $location,
      tag      => $target,
      ( $fcast ? (forecast => 1, days => 3) : () ),
    );
  }
}


sub pwx_error {
  my $err = $_[ARG0];

  my $status = $err->status;
  my $req    = $err->request;

  if ($req->tag) {
    my $chan = $req->tag;
    $_[HEAP]->irc->privmsg($chan => "Error: $err");
  }
  warn "Error: $err";
}

sub pwx_weather {
  my $res = $_[ARG0];

  my $place = $res->name;

  my $tempf = $res->temp_f;
  my $tempc = $res->temp_c;
  my $humid = $res->humidity;

  my $wind    = $res->wind_speed_mph;
  my $gust    = $res->wind_gust_mph;
  my $winddir = $res->wind_direction;
  
  my $terse   = $res->conditions_terse;
  my $verbose = $res->conditions_verbose;

  my $hms = $res->dt->hms;

  my $str = "$place at ${hms}UTC: ${tempf}F/${tempc}C";
  $str .= " and ${humid}% humidity;";
  $str .= " wind is ${wind}mph $winddir";
  $str .= " gusting to ${gust}mph" if $gust;
  $str .= ". Current conditions: ${terse}: $verbose";

  my $chan = $res->request->tag;
  $_[HEAP]->irc->privmsg($chan => $str);
}


sub pwx_forecast {
  my $res = $_[ARG0];

  my $place = $res->name;

  my $chan = $res->request->tag;

  $_[HEAP]->irc->privmsg($chan =>
    "Forecast for $place ->"
  );

  my $itr = $res->iter;
  while (my $day = $itr->()) {
    my $date = $day->dt->day_name;
    
    my $temp_hi_f = $day->temp_max_f;
    my $temp_lo_f = $day->temp_min_f;
    my $temp_hi_c = $day->temp_max_c;
    my $temp_lo_c = $day->temp_min_c;

    my $terse   = $day->conditions_terse;
    my $verbose = $day->conditions_verbose;

    my $wind    = $day->wind_speed_mph;
    my $winddir = $day->wind_direction;
    
    my $str = "${date}: High of ${temp_hi_f}F/${temp_hi_c}C";
    $str .= ", low of ${temp_lo_f}F/${temp_lo_c}C";
    $str .= ", wind $winddir at ${wind}mph";
    $str .= "; $terse: $verbose";
    
    $_[HEAP]->irc->privmsg($chan => $str);
  }
}

POE::Kernel->run



lib/POEx/Weather/OpenWeatherMap.pm
package POEx::Weather::OpenWeatherMap;
$POEx::Weather::OpenWeatherMap::VERSION = '0.002001';
use Defaults::Modern;

use POE 'Component::Client::HTTP';

use Weather::OpenWeatherMap::Cache;
use Weather::OpenWeatherMap::Error;

use Weather::OpenWeatherMap::Request;
use Weather::OpenWeatherMap::Request::Current;
use Weather::OpenWeatherMap::Request::Forecast;
use Weather::OpenWeatherMap::Request::Find;

use Weather::OpenWeatherMap::Result;
use Weather::OpenWeatherMap::Result::Current;
use Weather::OpenWeatherMap::Result::Forecast;
use Weather::OpenWeatherMap::Request::Find;


use Moo; use MooX::late;
with 'MooX::Role::POE::Emitter';


has api_key => (
  lazy        => 1,
  is          => 'ro',
  isa         => Str,
  predicate   => 1,
  builder     => sub { '' },
);

has _ua_alias => (
  lazy        => 1,
  init_arg    => 'ua_alias',
  is          => 'ro',
  isa         => Str,
  builder     => sub {
    my ($self) = @_;
    $self->alias ? $self->alias . 'UA'
      : confess "Cannot build ua_alias; emitter not running"
  },
);


has cache => (
  lazy        => 1,
  is          => 'ro',
  isa         => Bool,
  builder     => sub { 0 },
);

has cache_dir => (
  lazy        => 1,
  is          => 'ro',
  isa         => Maybe[Str],
  builder     => sub { undef },
);

has cache_expiry => (
  lazy        => 1,
  is          => 'ro',
  isa         => Maybe[StrictNum],
  builder     => sub { undef },
);

has _cache => (
  lazy        => 1,
  is          => 'ro',
  isa         => InstanceOf['Weather::OpenWeatherMap::Cache'],
  builder     => sub {
    my ($self) = @_;
    Weather::OpenWeatherMap::Cache->new(
      maybe dir    => $self->cache_dir,
      maybe expiry => $self->cache_expiry,
    )
  },
);


method start {
  $self->set_object_states(
    [
      $self => +{
        emitter_started  => 'mxrp_emitter_started',
        emitter_stopped  => 'mxrp_emitter_stopped',

        get_weather        => 'ext_get_weather',
        ext_http_response  => 'ext_http_response',
      },

      ( $self->has_object_states ? $self->object_states->all : () ),
    ]
  );
  $self->_start_emitter
}

method stop { $self->_shutdown_emitter }

method _emit_error (@params) {
  my $err = Weather::OpenWeatherMap::Error->new(@params);
  $self->emit( error => $err );
  $err
}


sub mxrp_emitter_started {
#  my ($kernel, $self) = @_[KERNEL, OBJECT];
}

sub mxrp_emitter_stopped {
  my ($kernel, $self) = @_[KERNEL, OBJECT];
  $kernel->call( $self->_ua_alias => 'shutdown' )
    if $kernel->alias_resolve( $self->_ua_alias );
}

method _result_type (Object $my_request) {
  state $base = 'Weather::OpenWeatherMap::Request::';
  my ($type, $event);
  CLASS: {
    if ($my_request->isa($base.'Current')) {
      $type  = 'Current';
      $event = 'weather';
      last CLASS
    }
    
    if ($my_request->isa($base.'Forecast')) {
      $type  = 'Forecast';
      $event = 'forecast';
      last CLASS
    }

    confess "Unknown request type: $my_request"
  }
  wantarray ? ($type, $event) : $type
}

method get_weather (@params) { $self->yield(get_weather => @params) }

sub ext_get_weather {
  my $self = $_[OBJECT];

  my %args = @_[ARG0 .. $#_];

  my $location = $args{location};
  unless ($location) {
    warn "Missing 'location =>' in query\n";
    my $fake_req = Weather::OpenWeatherMap::Request->new_for(
      Current =>
        tag      => $args{tag},
        location => '',
    );
    $self->_emit_error(
      source  => 'internal',
      request => $fake_req,
      status  => "Missing 'location =>' in query",
    );
    return
  }

  my $type = 
      delete $args{forecast} ? 'Forecast' 
    : delete $args{find}     ? 'Find'
    : 'Current';

  my $my_request = Weather::OpenWeatherMap::Request->new_for(
    $type =>
      ( 
        $self->has_api_key && length $self->api_key ?
          (api_key => $self->api_key) : () 
      ),
      %args
  );

  if ($self->cache && (my $cached = $self->_cache->retrieve($my_request)) ) {
    my $obj = $cached->object;
    my (undef, $event) = $self->_result_type($my_request);
    $self->emit( $event => $obj );
    return $obj
  }

  $self->_issue_http_request($my_request);
  ()
}


method _issue_http_request (Object $my_request) {
  unless ( $poe_kernel->alias_resolve($self->_ua_alias) ) {
    POE::Component::Client::HTTP->spawn(
      Alias           => $self->_ua_alias,
      FollowRedirects => 2,
    )
  }

  $poe_kernel->post( $self->_ua_alias => request => ext_http_response =>
    $my_request->http_request,
    $my_request
  );
}


sub ext_http_response {
  my $self = $_[OBJECT];

  my (undef, $my_request) = @{ $_[ARG0] };
  my ($http_response)     = @{ $_[ARG1] };

  unless ($http_response->is_success) {
    $self->_emit_error(
      source  => 'http',
      request => $my_request,
      status  => $http_response->status_line,
    );
    return
  }

  my ($type, $event) = $self->_result_type($my_request);

  my $content = $http_response->content;
  my $my_response = Weather::OpenWeatherMap::Result->new_for(
    $type =>
      request => $my_request,
      json    => $content,
  );
  
  unless ($my_response->is_success) {
    my $code = $my_response->response_code;
    $self->_emit_error(
      source  => 'api',
      request => $my_request,
      status  => "$code: ".$my_response->error,
    );
    return
  }

  $self->_cache->cache($my_response) if $self->cache;

  $self->emit( 
    $event => $my_response 
  );
}

1;


=pod

=for Pod::Coverage ext_\w+ mxrp_\w+ has_\w+

=head1 NAME

POEx::Weather::OpenWeatherMap - POE-enabled OpenWeatherMap client

=head1 SYNOPSIS

  use POE;
  use POEx::Weather::OpenWeatherMap;

  # An API key can be obtained (free) at http://www.openweathermap.org:
  my $api_key = 'foo';

  POE::Session->create(
    package_states => [
      main => [qw/
        _start
        
        pwx_error
        pwx_weather
        pwx_forecast
      /],
    ],
  );

  sub _start {
    my ($kernel, $heap) = @_[KERNEL, HEAP];
    
    # Create, store, and start emitter:
    my $wx = POEx::Weather::OpenWeatherMap->new(
      api_key      => $api_key,
      event_prefix => 'pwx_',
    );

    $heap->{wx} = $wx;
    $wx->start;

    ## An example request:
    $wx->get_weather(
      location => 'Manchester, NH',
      tag      => 'mytag',
    );
  }

  sub pwx_error {
    my $err = $_[ARG0];
    my $status  = $err->status;
    my $request = $err->request;
    # ... do something with error ...
    warn "Error! ($status)";
  }

  sub pwx_weather {
    my $result = $_[ARG0];

    my $tag = $result->request->tag;

    my $place = $result->name;

    my $tempf = $result->temp_f;
    my $conditions = $result->conditions_verbose;
    # (see Weather::OpenWeatherMap::Result::Current for a method list)
    # ...
  }

  sub pwx_forecast {
    my $result = $_[ARG0];

    my $place = $result->name;

    my $itr = $result->iter;
    while (my $day = $itr->()) {
      my $date = $day->dt->mdy;
      my $temp_hi = $day->temp_max_f;
      my $temp_lo = $day->temp_min_f;
      # (see Weather::OpenWeatherMap::Result::Forecast)
      # ...
    }
  }

  POE::Kernel->run;

=head1 DESCRIPTION

A POE-enabled interface to OpenWeatherMap (L<http://www.openweathermap.org>),
providing an object-oriented asynchronous interface to current & forecast
weather conditions for a given city, latitude/longitude, or OpenWeatherMap
city code.

This is really just an asynchronous counterpart to L<Weather::OpenWeatherMap>;
look there for documentation regarding Request & Result objects.

This an event emitter that consumes L<MooX::Role::POE::Emitter>; look there
for documentation on composed methods. See L<http://www.openweathermap.org>
for more on OpenWeatherMap itself.

=head2 ATTRIBUTES

=head3 api_key

Your L<OpenWeatherMap|http://www.openweathermap.org> API key.

(See L<http://www.openweathermap.org/api> to register for free.)

=head3 cache

A boolean value indicating whether successful results should be cached to disk
via L<Weather::OpenWeatherMap::Cache>.

Defaults to false. This may change in a future release.

=head3 cache_dir

The directory in which cache files are saved. The default may be fine; see
L<Weather::OpenWeatherMap::Cache>.

=head3 cache_expiry

The duration (in seconds) for which cache files are considered valid. The
default may be fine; see L<Weather::OpenWeatherMap::Cache>.

=head2 METHODS

=head3 start

Start our session.

Must be called before events will be received or emitted.

=head3 stop

Stop our session, shutting down the emitter and user agent (which will cancel
pending requests).

=head3 get_weather

  $wx->get_weather(
    # 'location =>' is mandatory.
    #  These are all valid location strings:
    #  By name:
    #   'Manchester, NH'
    #   'London, UK'
    #  By OpenWeatherMap city code:
    #   5089178
    #  By latitude/longitude:
    #   'lat 42, long -71'
    location => 'Manchester, NH',

    # Set 'forecast => 1' to get the forecast,
    # omit or set to false for current weather:
    forecast => 1,

    # If 'forecast' is true, you can specify the number of days to fetch
    # (up to 14):
    days => 3,

    # Optional tag for identifying the response to this request:
    tag  => 'foo',
  );

Request a weather report for the given C<< location => >>.

The location can be a 'City, State' or 'City, Country' string, an
L<OpenWeatherMap|http://www.openweathermap.org/> city code, or a 'lat X, long
Y' string.

Requests the current weather by default (see
L<Weather::OpenWeatherMap::Request::Current>).

If passed C<< forecast => 1 >>, requests a weather forecast (see
L<Weather::OpenWeatherMap::Request::Forecast>), in which case C<< days
=> $count >> can be specified (up to 14).

If passed C<< find => 1 >>, requests search results for a given location name
or latitude & longitude; see L<Weather::OpenWeatherMap::Request::Find>.

An optional C<< tag => >> can be specified to identify the response when it
comes in.

Any extra arguments are passed to the constructor for the appropriate Request
subclass; see L<Weather::OpenWeatherMap::Request>.

The request is made asynchronously and a response (or error) emitted when it
is available; see L</EMITTED EVENTS>. There is no useful return value.

=head2 RECEIVED EVENTS

=head3 get_weather

  $poe_kernel->post( $wx->session_id =>
    get_weather =>
      location => 'Manchester, NH',
      tag      => 'foo',
  );

POE interface to the L</get_weather> method; see L</METHODS> for available
options.

=head2 EMITTED EVENTS

=head3 error

Emitted when an error occurs; this may be an internal error, an HTTP error,
or an error reported by the OpenWeatherMap API.

C<$_[ARG0]> is a L<Weather::OpenWeatherMap::Error> object.

=head3 weather

Emitted when a request for the current weather has been successfully processed.

C<$_[ARG0]> is a L<Weather::OpenWeatherMap::Result::Current> object; see
that module's documentation for details on retrieving weather information.

=head3 forecast

Emitted when a request for a weather forecast has been successfully processed.

C<$_[ARG0]> is a L<Weather::OpenWeatherMap::Result::Forecast> object;
see that module's documentation for details on retrieving per-day forecasts
(L<Weather::OpenWeatherMap::Result::Forecast::Day> objects).

=head1 SEE ALSO

L<Weather::OpenWeatherMap>

L<Weather::OpenWeatherMap::Error>

L<Weather::OpenWeatherMap::Result>

L<Weather::OpenWeatherMap::Result::Current>

L<Weather::OpenWeatherMap::Result::Forecast>

L<Weather::OpenWeatherMap::Request>

L<Weather::OpenWeatherMap::Request::Current>

L<Weather::OpenWeatherMap::Request::Forecast>

The C<examples/> directory of this distribution.

=head1 AUTHOR

Jon Portnoy <avenj@cobaltirc.org>

=cut

# vim: ts=2 sw=2 et sts=2 ft=perl
